src/strings/ops.c

```diff
@@ -1693,13 +1718,27 @@ MVMString * MVM_string_join(MVMThreadContext *tc, MVMString *separator, MVMObjec
     }
     result->body.num_graphs = total_graphs;
 
-    /* If we just collect all the things as strands, are we within bounds, and
-     * will be come out ahead? */
-    if (total_strands < MVM_STRING_MAX_STRANDS && 16 <= total_graphs / total_strands) {
-        /* XXX TODO: Implement this, conditionalize branch thing below. */
+    /* If the separator and pieces are all strands of the same type,
+     * and there are enough strands of a large enough size, just copy
+     * them directly. */
+    if (total_strands <  MVM_STRING_MAX_STRANDS &&
+            16        <= total_graphs / total_strands &&
+            all_strands_of_same_type) {
+        MVMuint16 offset = 0;
+        result->body.storage_type    = MVM_STRING_STRAND;
+        result->body.storage.strands = allocate_strands(tc, total_strands);
+        result->body.num_strands     = total_strands;
+        for (i = 0; i < num_pieces; i++) {
+            MVMString *piece = pieces[i];
+            if (0 < i) {
+                copy_strands(tc, separator, 0, result, offset, separator->body.num_strands);
+                offset += separator->body.num_strands;
+            }
+            copy_strands(tc, piece, 0, result, offset, piece->body.num_strands);
+            offset += piece->body.num_strands;
```

Comment:
This code doesn't check if the result is NFG stable or not, which means it'll produce wrong results if two of the things being joined would form a single grapheme.

Reply:
I wondered about this, but couldn't craft an example where the behavior after my change did anything different than the code before. I tried various combinations of joining elements that had strands of just combining characters, just combining characters as the separator, etc. Also, it does pass a spectest. Either the collapsing is correctly happening elsewhere, my ad-hoc test cases weren't actually testing the right things, or we should add some tests for this to roast. Do you have an example I can test?